Serialize whether VarDecls are top-level

adrian-prantl opened the pull request:
The current way that VarDecl::isLazilyInitializedGlobal() is implemented does
not work in the debugger, since the DeclContext of all VarDecls are deserialized
Swift modules. By adding a bit to the VarDecl we can recover the fact that a
VarDecl was in fact a global even in the debugger.

rdar://problem/58939370

xedin committed:
[CSFix] Add a fix to add a missing qualifier to shadowed top-level name references

xedin committed:
[Diagnostics] Port name shadowing diagnostics

Diagnose an attempt to reference a top-level name shadowed by
a local member e.g.

```swift
extension Sequence {
  func test() -> Int {
    return max(1, 2)
  }
}
```

Here `min` refers to a global function `min<T>(_: T, _: T)` in `Swift`
module and can only be accessed by adding `Swift.` to it, because `Sequence`
has a member named `min` which accepts a single argument.

xedin committed:
[ConstraintSystem] Use new fix/diagnostic for name shadowing

Stop filtering outer overload choices while trying to pre-check
expression, instead have it always fetch those and use new
fix to only attempt them in diagnostic mode (unless it's min/max
situation with conditional conformances).

xedin committed:
[CSDiag] NFC: Remove obsolete name shadowing diagnostics

xedin committed:
[Diagnostics] NFC: Fix name shadowing diagnostic comment to refer to `max`

xedin committed:
[TypeChecker] Add types nested in protocol to lookup results

Although such functionality is not yet supported we have to
mirror AST lookup and add such members into results, otherwise
there is a risk of inner/outer results mismatch.

adrian-prantl committed:
Serialize whether a VarDecl is a top-level global.

The current way that VarDecl::isLazilyInitializedGlobal() is implemented does
not work in the debugger, since the DeclContext of all VarDecls are deserialized
Swift modules. By adding a bit to the VarDecl we can recover the fact that a
VarDecl was in fact a global even in the debugger.

<rdar://problem/58939370>

adrian-prantl commented:
Any pointers for how to test this would be appreciated.

xedin commented:
I can pull this change into my branch and run full suite test to see whether it fixes the problem.

xedin commented:
@adrian-prantl I have `am`ed your commit to https://github.com/apple/swift/pull/29024 and started full test.

adrian-prantl commented on lib/SILGen/SILGenGlobalVariable.cpp:
```diff
@@ -69,7 +69,8 @@ ManagedValue
 
-  if (var->isLazilyInitializedGlobal()) {
+  // In the debugger, no variables are lazy initialized.
+  if (!SGM.SwiftModule->getDebugClient() && var->isLazilyInitializedGlobal()) {
```
Sorry, I accidentally left this is.

xedin replied:
No worries, I can re-apply and test :)

xedin commented:
I have re-applied the changes and restarted testing.

adrian-prantl commented:
test with https://github.com/apple/llvm-project/pull/674
@swift-ci test

adrian-prantl commented:
@xedin You'll also need https://github.com/apple/llvm-project/pull/674 for the LLDB tests to succeed. This PR tests whether it also works *without* your PR. If yes I will land this, cherry-pick both to master-rebranch and master-next and then you should be able to merge your PR.

xedin commented:
Sounds good! Going to keep an eye on this PR.

adrian-prantl commented:
test with https://github.com/apple/llvm-project/pull/674
@swift-ci test

adrian-prantl commented:
test with https://github.com/apple/llvm-project/pull/674
@swift-ci smoke test

adrian-prantl committed:
Merge remote-tracking branch 'xedin/diagnose-shadowing'

adrian-prantl commented:
test with https://github.com/apple/llvm-project/pull/674
@swift-ci smoke test

adrian-prantl committed:
Compile libraries in testcases with -parse-as-library (NFC)

This is in preparation to a change in serialization of global variables where
this detail will matter.

adrian-prantl commented:
test with https://github.com/apple/llvm-project/pull/674
@swift-ci test

adrian-prantl commented:
test with https://github.com/apple/llvm-project/pull/674
@swift-ci test

slavapestov commented on lib/AST/Decl.cpp:
```diff
@@ -5410,11 +5411,7 @@ bool VarDecl::isLazilyInitializedGlobal() const {
 
   // Top-level global variables in the main source file and in the REPL are not
```
Are the above checks (hasClangNode() and isDebuggerVar()) needed at all then? The bit should not be set in those cases either.

slavapestov reviewed (commented):
This looks good, but it appears you've cherry-picked @xedin's changes into this PR too. Can we land the two sets of changes separately?

xedin commented:
@slavapestov We actually going to do it vice versa - I have cherry-picked Adrian's changes to my PR to make sure everything works correctly, currently waiting on full performance suite to complete and we are going to land my PR together with llvm side.

slavapestov commented:
@xedin Ok, sure. In that case, can you simplify VarDecl::isLazilyInitializedGlobal() in your PR then?

xedin commented:
@slavapestov Sure! I'm going to let full test finish and commit the simplification to my PR.

adrian-prantl commented:
Abandoning this in favor of @xedin's all-in-one PR.
